internal/tmpl/tree_test.go

```diff
@@ -160,7 +169,7 @@ func TestNonEmptyRangeSerialization(t *testing.T) {
 	if n != int64(buf.Len()) {
 		t.Fatalf("wrote %d but tracked %d", buf.Len(), n)
 	}
-	const want = `{"0":"abc","1":{"d":[["1"],["2"],["3"]],"s":["x is ","."]},"s":["","def",""]}`
+	const want = `{"0":"abc","1":{"d":[["1"],["2"],["3"]],"s":["x is ",".","",""]},"s":["","def",""]}`
```

Comment:
This is one of the harder cases.  Here we have a range of dynamics in the `"d"` array.  Each entry in the `"d"` array is also an array.  These "dynamic entry arrays" are of length 1 while the dynamic array itself has a length of 3.  In this case, the statics array (`"s"`) should be of length 2 which is the length of (all) the "dynamic entry array" not the length of the dynamic array.  In other words, the 2 additional empty strings is incorrect and will result in 2 "undefined" being appended to the html when rendered.

Reply:
i see. ok, i've cleaned up and reorganized a bunch of this code. i'm still not sure i've got it right, because there's no spec to work against, but this at least passes all the existing tests and is a lot more internally consistent.